ams/server/src/main/java/com/netease/arctic/server/manager/FlinkOptimizerContainer.java

```diff
@@ -371,6 +438,93 @@ private String buildReleaseKubernetesCommand(Resource resource) {
         this.flinkHome, options);
   }
 
+  /** Submit the OptimizerJob to the flink cluster through Flink RestClient. */
+  protected JobID runFlinkOptimizerJob(Resource resource, Map<String, String> flinkRestConf)
+      throws Exception {
+    try (RestClusterClient<String> client = FlinkClientUtil.getRestClusterClient(flinkRestConf)) {
+      ClusterOverviewWithVersion overviewWithVersion =
+          client.getClusterOverview().get(flinkClientTimeoutSecond, TimeUnit.SECONDS);
+      LOG.debug("cluster overview: {}", overviewWithVersion);
+      int numSlotsAvailable = overviewWithVersion.getNumSlotsAvailable();
+      if (numSlotsAvailable < resource.getThreadCount()) {
+        throw new RuntimeException("No enough slots available in the cluster.");
+      }
+      File jarFile = new File(jobUri);
+      Preconditions.checkArgument(
+          jarFile.exists(), String.format("The jar file %s not exists", jarFile.getAbsolutePath()));
+
+      URL url = new URL(client.getWebInterfaceURL());
+      return runJar(
+          uploadJar(jarFile, url.getHost(), url.getPort(), flinkRestConf), flinkRestConf, resource);
```

Comment:
need to upload the JAR every time, or will you take the previous JAR?

Reply:
No, before uploading, the  flink Rest api will be called to check whether the optmizer jar package exists in the current cluster. The current stage is to determine whether it already exists through the file name. This may cause problems (the name is the same but the content is different), but I I think this is enough in most cases. If the jar package is indeed updated but the jar package name is not updated, just go to flink web ui and delete it manually.